crates/key-server/src/mvr.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c), Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::errors::InternalError;
+use crate::errors::InternalError::{Failure, InvalidMVRName};
+use move_core_types::account_address::AccountAddress;
+use move_core_types::identifier::Identifier;
+use move_core_types::language_storage::StructTag;
+use serde_json::json;
+use std::str::FromStr;
+use sui_sdk::rpc_types::SuiParsedData;
+use sui_sdk::SuiClient;
+use sui_types::base_types::ObjectID;
+use sui_types::dynamic_field::DynamicFieldName;
+use sui_types::TypeTag;
+
+const MVR_REGISTRY: &str = "0xe8417c530cde59eddf6dfb760e8a0e3e2c6f17c69ddaab5a73dd6a6e65fc463b";
+const MVR_CORE: &str = "0x62c1f5b1cb9e3bfc3dd1f73c95066487b662048a6358eabdbf67f6cdeca6db4b";
+
+/// Given an MVR name, look up the package it points to.
+pub(crate) async fn mvr_forward_resolution(
+    client: &SuiClient,
+    mvr_name: &str,
+) -> Result<ObjectID, InternalError> {
+    let dynamic_field = client
+        .read_api()
+        .get_dynamic_field_object(
+            ObjectID::from_str(MVR_REGISTRY).unwrap(),
+            dynamic_field_name(mvr_name)?,
+        )
+        .await
+        .map_err(|_| InvalidMVRName)?
+        .data
+        .ok_or(Failure)?
+        .content
+        .ok_or(Failure)?;
+
+    let package_address_as_str = match dynamic_field {
+        SuiParsedData::MoveObject(obj) => obj.fields.to_json_value()["value"]["app_info"]
+            ["package_address"]
+            .as_str()
+            .ok_or(Failure)?
+            .to_string(),
+        _ => return Err(Failure),
+    };
+
+    ObjectID::from_str(&package_address_as_str).map_err(|_| Failure)
+}
+
+fn dynamic_field_name(mvr_name: &str) -> Result<DynamicFieldName, InternalError> {
+    let parsed_name =
+        mvr_types::name::VersionedName::from_str(mvr_name).map_err(|_| InvalidMVRName)?;
+    if parsed_name.version.is_some() {
```

Comment:
what does it mean?

Reply:
An MVR name can have the form `@mysten/kiosk/3` to require that version 3 is used, but this check ensures that only names without the version can be used.